plugins/modules/identity/keycloak/keycloak_user.py

```diff
@@ -0,0 +1,527 @@
+        module.fail_json(msg=str(e))
+
+    kc = KeycloakAPI(module, connection_header)
+
+    realm = module.params.get('realm')
+    state = module.params.get('state')
+    username = module.params.get('username')
+    attributes = module.params.get('attributes')
+
+    # attributes in Keycloak have their values returned as lists
+    # via the API. attributes is a dict, so we'll transparently convert
+    # the values to lists.
+    if attributes is not None:
+        for key, val in module.params['attributes'].items():
+            module.params['attributes'][key] = [
+                val] if not isinstance(val, list) else val
+
+    # Filter and map the parameters names that apply to the user
+    user_params = [x for x in module.params
+                   if x not in list(keycloak_argument_spec().keys()) + ['state', 'realm'] and module.params.get(x) is not None]
+
+    found_users = kc.get_users_by_username(username, realm=realm)
+
+    if found_users is None or found_users == []:
+        before_user = {}
+    else:
+        before_user = found_users[0]
+
+    # Build a proposed changeset from parameters given to this module
+    changeset = {}
+
+    for param in user_params:
+        new_param_value = module.params.get(param)
+        old_value = before_user[param] if param in before_user else None
+        if new_param_value != old_value:
+            changeset[camel(param)] = new_param_value
+
+    # Prepare the desired values using the existing values (non-existence results in a dict that is save to use as a basis)
+    desired_user = before_user.copy()
+    desired_user.update(changeset)
+
+    # Cater for when it doesn't exist (an empty dict)
+    if not before_user:
+        if state == 'absent':
+            # Do nothing and exit
+            if module._diff:
+                result['diff'] = dict(before='', after='')
+            result['changed'] = False
+            result['end_state'] = {}
+            result['user'] = result['end_state']
+            result['msg'] = 'User does not exist; doing nothing.'
+            module.exit_json(**result)
+
+        # Process a creation
+        result['changed'] = True
+
+        if module._diff:
+            result['diff'] = dict(before='', after=desired_user)
+
+        if module.check_mode:
+            module.exit_json(**result)
+
+        # create it
+        kc.create_user(desired_user, realm=realm)
+
+        after_user = kc.get_users_by_username(username, realm=realm)[0]
+
+        result['end_state'] = after_user
+        result['user'] = result['end_state']
+
+        result['msg'] = 'User {username} has been created with ID {id}'.format(username=after_user['username'],
+                                                                               id=after_user['id'])
+        module.exit_json(**result)
+
+    else:
+        if state == 'present':
+            # Process an update
+
+            # no changes
+            if desired_user == before_user:
```

Comment:
@fynncfchen First, thank you for the module.I have been testing it against both keycloak 14 and 16.1 during the day and I think I found a bug.Not sure how to fix it sadly. 

Consider the following snippet:

```yaml
- keycloak_user:
    auth_client_id: admin-cli
    auth_keycloak_url: ...
    token: ...
    username: test
    realm: master
    credentials:
      - type: password
        temporary: false
        value: test
    state: present
```

During the first execution the user test will be correctly created (“User test has been created with ID blabla”).
On subsequent runs instead of doing nothing the user is updated every time.
In these cases the variables `before_user` and `after_user` are strictly identical.
The difference comes with `desired_users` which contains a key `credentials` while `before_user` and `after_user` don't.

Obviously keycloak is not going to return clear text passwords to compare against and simply comparing `before_user` and `after_user` is not going to catch a password change.